azkaban-web-server/src/main/java/azkaban/flowtrigger/db/JdbcFlowTriggerInstanceLoaderImpl.java

```diff
@@ -0,0 +1,466 @@
+      this.dbOperator.transaction(tran);
+    } catch (final SQLException ex) {
+      handleSQLException(ex);
+    }
+  }
+
+  @Override
+  public void uploadTriggerInstance(final TriggerInstance triggerInst) {
+    final SQLTransaction<Integer> insertTrigger = transOperator -> {
+      for (final DependencyInstance depInst : triggerInst.getDepInstances()) {
+        transOperator
+            .update(INSERT_DEPENDENCY, triggerInst.getId(), depInst.getDepName(),
+                depInst.getStartTime(), depInst.getEndTime(), depInst.getStatus().ordinal(),
+                depInst.getCancellationCause().ordinal(),
+                triggerInst.getProject().getId(),
+                triggerInst.getProject().getVersion(),
+                triggerInst.getFlowId(),
+                triggerInst.getFlowVersion(),
+                FlowUtils.toJson(triggerInst.getProject()),
+                triggerInst.getFlowExecId());
+      }
+      return null;
+    };
+
+    executeTransaction(insertTrigger);
+  }
+
+  @Override
+  public void updateDependencyExecutionStatus(final DependencyInstance depInst) {
+    executeUpdate(UPDATE_DEPENDENCY_STATUS_ENDTIME_AND_CANCELLEATION_CAUSE, depInst.getStatus()
+            .ordinal(),
+        depInst.getEndTime(), depInst.getCancellationCause().ordinal(),
+        depInst.getTriggerInstance().getId(),
+        depInst.getDepName());
+  }
+
+  /**
+   * Retrieve recently finished trigger instances, but flow trigger properties are not populated
+   * into the returned trigger instances for efficiency. Flow trigger properties will be
+   * retrieved only on request time.
+   */
+  @Override
+  public Collection<TriggerInstance> getRecentlyFinished(final int limit) {
+    final String query = String.format(SELECT_RECENTLY_FINISHED, limit);
+    try {
+      return this.dbOperator.query(query, new TriggerInstanceHandler());
+    } catch (final SQLException ex) {
+      handleSQLException(ex);
+    }
+    return Collections.emptyList();
+  }
+
+  /**
+   * Retrieve a trigger instance given an instance id. Flow trigger properties will also be
+   * populated into the returned trigger instance.
+   */
+  @Override
+  public TriggerInstance getTriggerInstanceById(final String triggerInstanceId) {
+    TriggerInstance triggerInstance = null;
+    try {
+      final Collection<TriggerInstance> res = this.dbOperator
+          .query(SELECT_EXECUTIONS_BY_INSTANCE_ID, new TriggerInstanceHandler(), triggerInstanceId);
+      triggerInstance = !res.isEmpty() ? res.iterator().next() : null;
+    } catch (final SQLException ex) {
+      handleSQLException(ex);
+    }
+    if (triggerInstance != null) {
+      final int projectId = triggerInstance.getProject().getId();
+      final int projectVersion = triggerInstance.getProject().getVersion();
+      final String flowFileName = triggerInstance.getFlowId() + ".flow";
+      final int flowVersion = triggerInstance.getFlowVersion();
+      final File tempDir = Files.createTempDir();
```

Comment:
I don't understand, why are we creating a temp dir then deleting it?

Reply:
it's required by the API. Though I think the API can be improved, but the temp dir is used to store the downloaded file and it's caller's responsibility to delete the dir. cc @jamiesjc